### ISSUE TITLE
"HTTP request to the remote WebDriver timed out in 60 seconds" when commandTimeout passed to RemoteWebDriver constructor c#

### ISSUE THREAD

**Issue by jburns24** [SeleniumHQ/selenium#4942]
## Meta -
OS:  Windows 2012R2

Selenium Version:  webdriver and support 3.6.0

Browser:  chrome, firefox, internet explorer

Browser Version:  

Chrome 61.0.3163.100
IE 11
Firefox 56.0.1

## Expected Behavior -
When registering a remote webdriver like so 
```
InternetExplorerOptions ieOptions = new InternetExplorerOptions();
ieOptions.EnsureCleanSession = true;
ieOptions.RequireWindowFocus = true;
ieOptions.IgnoreZoomLevel = true;
ieOptions.BrowserAttachTimeout = TimeSpan.FromSeconds(60);
ICapabilities ieCaps = ieOptions.ToCapabilities();
Driver = new RemoteWebDriver(new Uri("http://chi-grid-hub1.corp.auctiva.com:4444/wd/hub"), ieCaps, TimeSpan.FromMinutes(2));
```
that WebDriverTimeouts are thrown only when command responses take longer than 2 min. 

## Actual Behavior -
Tests are periodically failing after no response is received in 60 seconds. Here is a screenshot of the exception 
![capture](https://user-images.githubusercontent.com/19497855/31966354-4d9d3254-b8bf-11e7-8480-2291d4394d3f.PNG)

I would expect this timeout to only be triggered after 2 min has passed. When debugging the driver I see both `DefaultCommandTimeout` and 'serverResponseTimeout' set 
![capture](https://user-images.githubusercontent.com/19497855/31966563-ec7c701a-b8bf-11e7-8990-b257cf7ba475.PNG)
Seems to me that DefaultCommandTimeout is being used regardless of serverResonseTimeout, althought my only 'evidence' for this is me knowing I do not explicitly set any timeouts to 60 seconds and the behavior, of a .Click() being executed and then in CreateResonse() a failure after 60 seconds. 

## Steps to reproduce -
I do not have a test case to share being I experience this issue on internal environments. I do not know of a public site where actions take longer than 1 min to return a response. I am gonna try and get a case for you. 
Also to my knowledge I have insured that the application code is only setting this one timeout in the driver constructor, and my node and grid timeouts are default. I tried changing the `unregisterIfStillDownAfter` attribute in the node configuration but that didnt do change the failure, and if I based on the comments it seems this attribute is only used to unregister a node if it fails to register to the hub before the timeout.

**Comment by bitterpeanut** [comment]
serverResponseTimeout: The timeout within which the server must respond.
DefaultCommandTimeout: The default command timeout for HTTP requests in a RemoteWebDriver instance.

I'm a little unclear in what you are trying to communicate. In the beginning you state on "WebDriverTimeouts are thrown only when command responses take longer than 2 min." then towards the end you state "Seems to me that DefaultCommandTimeout is being used regardless of serverResonseTimeout". Those statements seem to contradict each other. Which one are you experiencing?

The two timeouts are for difference situations. 

The serverResponseTimeout is saying this number of seconds has passed and the driver hasn't heard anything back from the remote server in 2 minutes. (server timeout) Where as the DefaultCommandTimeout is saying it sent an event to the page and this many number of seconds has passed without a response from the page. (browser timeout)

You can increase the browser timeout of 60 seconds when you [setup your grid](https://github.com/SeleniumHQ/selenium/wiki/Grid2) node. 

> On the hub you can also set -browserTimeout 60 to make the maximum time a node is willing to hang inside the browser 60 seconds.

Not sure if this helps, but I hope it can solve the issue you are seeing.

**Comment by jburns24** [comment]
Those two things seemed different, though I had convinced myself they were the same when I was looking at the C# source code. Looking at the 3.6 source I see I was calling this constructor
![capture](https://user-images.githubusercontent.com/19497855/32000992-4414ca88-b94d-11e7-8b8d-6420037f5ebd.PNG)
This was then calling the `HttpCommandExecutor` constructor 
![capture](https://user-images.githubusercontent.com/19497855/32001063-7a3d2376-b94d-11e7-8edd-44c984f5dde1.PNG)

So I guess this is where my confusion starts. In the `RemoteWebDriver` constructor it seems that the `TimeSpan` you pass is used as the `commandTimout` (according to the comment). But then is passed to the `HttpCommandExecutor()` constructor which sets the `serverResponseTimeout`. You suggest overriding this with the hub config but when I look at my current hub config I have `browserTimeout` set to 0. Which according to the documentation should wait indefinitely. 
![capture](https://user-images.githubusercontent.com/19497855/32001261-1e3a16f0-b94e-11e7-969f-2eee3221de5f.PNG)
So it seems to me that I am constructing the driver with the constructor that should set the CommandTimeout, on a grid configured to have a browserTimout of 0. Yet I am seeing test fail with the above exception "

But to clarify your original question I am seeing I am seeing the above exception thrown when commands take longer than 1 min to respond when I would expect them to fail if the response takes longer than 2 min. For some context, the use case I have is I am clicking a button that triggers an expensive SQL query that can take an upwards of 1 min to execute. 

So the stack trace shows the exception comes from `HttpWebRequest.GetResponse()` looking at how that gets implemented in `HttpCommandExecutor` I see this 
![capture](https://user-images.githubusercontent.com/19497855/32006254-1ee38cf0-b95b-11e7-9289-3473ddfe0741.PNG)

Which looks like is using the `serverResponseTimeout`. I can see in debugging that after initializing the driver the `serverResponseTimeout` is set to 2 min but based on the formatting of this exception message and that the second object is `this.serverResponseTimeout.TotalSeconds` which based on the exception message I actually get seems to be 60 seconds.

**Comment by kotprav** [comment]
Our team sees the same problem with IE driver.

**Comment by diemol** [comment]
Yes, this was a bug present in several releases and I believe it was fixed in 3.12. 
To be safe, please try again with the latest release of Grid 3.x and in case the issue is still happening, please open a new issue.
